kymatio/scattering2d/frontend/torch_frontend.py

```diff
@@ -0,0 +1,181 @@
+        pre_pad : boolean
+            Controls the padding: if set to False, a symmetric padding is applied
+            on the signal. If set to true, the software will assume the signal was
+            padded externally.
+        Psi : dictionary
+            Contains the wavelets filters at all resolutions. See
+            filter_bank.filter_bank for an exact description.
+        Phi : dictionary
+            Contains the low-pass filters at all resolutions. See
+            filter_bank.filter_bank for an exact description.
+        M_padded, N_padded : int
+             Spatial support of the padded input.
+
+        Notes
+        -----
+        The design of the filters is optimized for the value L = 8.
+
+        pre_pad is particularly useful when cropping bigger images because
+        this does not introduce border effects inherent to padding.
+
+        """
+    def __init__(self, J, shape, L=8, max_order=2, pre_pad=False):
+        super(Scattering2D, self).__init__(J, shape, max_order=max_order)
+        self.pre_pad, self.L = pre_pad, L
+        if 2 ** J > shape[0] or 2 ** J > shape[1]:
+            raise RuntimeError('The smallest dimension should be larger than 2^J')
+        self.build()
+
+    def build(self):
+        self.M, self.N = self.shape
+        self.modulus = Modulus()
+        self.M_padded, self.N_padded = compute_padding(self.M, self.N, self.J)
+        # pads equally on a given side if the amount of padding to add is an even number of pixels, otherwise it adds an extra pixel
+        self.pad = Pad([(self.N_padded - self.N) // 2, (self.N_padded - self.N + 1) // 2, (self.M_padded - self.M) // 2,
+                       (self.M_padded - self.M + 1) // 2], [self.N, self.M], pre_pad=self.pre_pad)
+        self.subsample_fourier = SubsampleFourier()
+        self.create_and_register_filters()
+
+    def create_and_register_filters(self):
+        """ This function run the filterbank function that
+            will create the filters as numpy array, and then, it
+            saves those arrays as module's buffers."""
+
+        # Create the filters
+        filters = filter_bank(self.M_padded, self.N_padded, self.J, self.L)
+        n = 0
+        self.phi, self.psi = filters['phi'], filters['psi']
+        for c, phi in self.phi.items():
+            if isinstance(c, int):
+                self.phi[c] = torch.from_numpy(self.phi[c])
+                self.register_buffer('tensor' + str(n), self.phi[c])
+                n += 1
+
+        for j in range(len(self.psi)):
+            for k, v in self.psi[j].items():
+                if isinstance(k, int):
+                    self.psi[j][k] = torch.from_numpy(v)
+                    self.register_buffer('tensor' + str(n), self.psi[j][k])
+                    n += 1
+
+    def scattering(self, input):
+        # each time scattering is run, one needs to make sure self.psi and self.phi point to
+        # the correct buffers
```

Comment:
Is this because of potential `.to` and `.cuda` calls in the meantime?

If so, wouldn't it be semantically more correct to overwrite `.cuda` and `.to` to first call `super().to/cuda` and then make the modification in the dict?

Reply:
NO! We should never handle those functions on our own, except for raising an error for types not supported yet by backends(eg, `.half()`)

Reply:
(also pytorch has some local optimization we do not wish to deal with)